### ISSUE TITLE
Optimize time between job creation and handling

### ISSUE THREAD

**Issue by npepinpe** [camunda/zeebe#11231]
**Description**

We would like to reduce the time between a job is created until it is received by a `JobHandler`. One idea is to explore a push-based approach, where workers can open long living streams and receive jobs as soon as they are available, without having to poll for them.

To reiterate, the primary goal is: **reduce the time between job creation and client handling**. It is not to create a push-based pipeline - this is simply a proposed solution.

As a first step, we'll investigate a push based approach. The first iteration will consist of getting feedback on the performance gain as early as possible. This means ignoring most failure paths, most edge cases, QoL features, etc.

```[tasklist]
### Prototype
- [x] #11232
- [x] #11233
```

```[tasklist]
### Organizational
- [x] Define scope for [alpha target](https://github.com/camunda/zeebe/issues/11231#issuecomment-1444514091)
- [x] Define scope for first production-ready target
- [x] Provide an estimated timeline of deliverables
```

The initial alpha target will implement end-to-end job pushing (from broker to client), with naive solution for the more complex problems (log persistence, failure handling, and flow control), and will only implement support for the Java client.

```[tasklist]
### Alpha
- [ ] #11706
- [ ] #11707
- [ ] #11711
- [ ] #12010
- [ ] #11713
- [x] Define interface for StreamMetrics in transport and provide concrete class in Broker. Follow up of #12010
- [ ] #12253
- [ ] #12385
- [ ] #12384
- [ ] #12386
- [ ] #12387
- [ ] #12388
- [ ] #12389
- [ ] #12390
- [ ] #11712
- [ ] #11708
- [ ] #11709
- [ ] #11710
- [ ] https://github.com/camunda/zeebe/issues/13281
- [ ] https://github.com/camunda/zeebe/issues/13061
- [ ] https://github.com/camunda/zeebe/issues/13049
- [ ] https://github.com/camunda/zeebe/issues/13048
- [ ] https://github.com/camunda/zeebe/issues/13038
- [ ] https://github.com/camunda/zeebe/issues/13037
- [ ] https://github.com/camunda/zeebe/issues/13036
- [ ] https://github.com/camunda/zeebe/issues/13465
- [ ] https://github.com/camunda/zeebe/issues/13516
- [ ] https://github.com/camunda/zeebe/issues/13846
- [ ] https://github.com/camunda/zeebe/issues/13914
- [ ] https://github.com/camunda/zeebe/issues/14033
```

```[tasklist]
### Production
- [x] Prototype client flow control solution
- [x] Allow clients to control job streaming flow/back pressure
- [x] Document client flow control solution
- [ ] https://github.com/camunda/zeebe/issues/13915
- [ ] #12029
```

### Timeline

We'll approach this in two phases. Alpha and production-ready. Target for the alpha scope is 8.3.0-alpha2 (so the June release). Target for production-ready would be 8.3 (so September).

**Comment by npepinpe** [comment]
A quick summary (you can read more in the prototype issue), the prototype was a success I think. There's still many issues to solve, but it's clear the approach is much faster:

- For the common case, you skip having to send an ACTIVATE command. This means removing a required commit from your hot path.
- For the common case, it removes the need to query every partition from the gateway, so you don't need to wait for each request.

It's however a more complex approach than simple polling, and we would likely want to continue supporting polling, at least for the foreseeable future, which means increasing the complexity of the application. That said, the performance is much, much faster for the happy path, so it's worth pursuing.

**Comment by npepinpe** [comment]
I've updated the issue with a tentative breakdown for the alpha target. Next week, me, @deepthidevaki, and @koevskinikola will fix the scope for the alpha and production-ready target and finalize a provisional timeline of deliverables. This will be a soft timeline to help us better organize our time, not a fixed one.

**Comment by npepinpe** [comment]
## Alpha scope

> **Note** The main goal of the first alpha is to provide the minimum fundamental parts of the feature on which we can iterate. The idea is that we cannot properly compare various solutions for flow control, failure handling, job routing, etc., without having the base pipeline in place.

Clients will be able to open so-called job streams. A job stream will target a specific job type, and will define a set of job activation properties identical to the existing job worker. A client job stream will consist of a long living unidirectional stream between the client and the gateway, on which the gateway will push `ActivatedJob` instances. **This ensures compatibility with the `JobHandler` interface and the existing `JobWorker` API**.

The gateway will aggregate clients by their activation properties. For the first client with logically equivalent activation properties, it will open a stream to each broker for this job type, passing these properties and a unique ID identifying the aggregated stream. Clients which open a stream after that, on this gateway, will not cause the gateway to send a request to the broker. On the last client to close a job stream of a given type, the gateway will send a request to the broker to close the stream for this type to itself.

> **Note** Both the gateway and the broker will use our membership service to detect membership changes, and add/remove job streams accordingly.

The broker will have a new top-level service: the job stream service. This consists of a shared job stream registry, an API request handler to add/remove streams from the registry, and a streaming client to push activated jobs to registered streams.

The job stream registry will provide a synchronous, thread-safe API to fetch a job stream, if any, for a given job type. **This must be synchronous since it will be accessed by the engine during command processing**. A job stream in the registry will consist of an association of a job type, a set of job activation properties, and a list of possible recipients (gateway + unique aggregated stream ID).

The API request handler will receive the gateway’s open/close stream requests, and add/ remove the gateway from the possible recipients for activated jobs of a given type. When the last recipient is removed from a job stream, it is removed from the registry.

When a job would be made available, the engine will query the registry for an available stream. If there is none, then the job is made activate-able, and the job type is broadcasted. **This is important for compatibility with the current long-polling approach.** If there is one, then the job will be activated immediately during the same command processing, using the stream’s job activation properties. Once activated, the job and its variables are handed over (as a side-effect) to a specific consumer which will then forward the job to the stream's gateway (including the job, its key, variables, and the unique aggregated stream ID).


In more details:

### Client

- The **Java** client can open a long living stream which can activate jobs and receive them. It can pass a **worker name**, **activation timeout**, and a set of **fetch variables**. The semantics for activation will be the exact same ones as that with the `ActivateJob` command. It will receive a stream of `ActivatedJob` instances, such that the results can be fed to a `JobHandler` that it shares with a `JobWorker`.
- Activation of existing jobs before the stream is created will be handled initially by the existing job worker/poller.
  - As a stretch goal, if possible, we can integrate it as an opt-in feature of the job worker, so that the stream can be used seamlessly with the worker (i.e. the same job handler is reused). As we don't have experimental features yet on the client-side, we should consider whether this is worth doing now or wait.
- The client will emit a metric of how many jobs it receives (note: if this is already achievable with standard gRPC metric interceptors, this can be omitted).
- If the stretch goal is met, then we should also implement https://github.com/camunda/zeebe/issues/4700

### Gateway

#### Stream management
- The gateway will implement a new `JobStream` API, which can handle long-living client streams. Streams will be aggregated by their activation properties, and the aggregated stream given a unique ID. 
  - When a new aggregated stream is created, the gateway will register it to all brokers.
  - When a new aggregated stream is removed, the gateway will remove it from all brokers.
  - When a new broker is added to the topology, the gateway will register each of its existing aggregated streams.
  - When the gateway is shutting down, it should broadcast a single, best-of effort request to all brokers to remove its streams.
  - When a client is detected as dead, if it is the last client for the aggregated stream, the gateway will notify all brokers to remove this stream.
- The gateway will emit metrics of how many clients are registered and how many unique streams there are.

#### Job proxy
- The gateway will implement a new internal endpoint to receive a job, a job key, and the aggregated stream unique ID.
  - Received jobs will be transformed into `ActivatedJob` and forwarded in a round-robin fashion to one of the client streams associated with the aggregated stream.
  - If there are no clients available, the gateway will yield the job back to its partition leader.
  - If the client is detected as dead when forwarding the job, the gateway will yield the job back to its partition leader.
- The gateway will emit metrics reporting how many jobs it receives per partition, the rate of reception, and how many jobs it forwards, and the rate of forwarding.

### Broker

#### Stream management

- The broker will implement a new API endpoint to manage aggregated job streams, where gateways can add and remove unique streams, and remove all streams associated to a gateway.
  - When a gateway is removed from the topology, all associated streams should be removed from the broker.

#### Job push

- The broker will expose via an interface the set of registered unique streams and their activation properties to the stream platform, such that the engine can activate jobs with these properties.
- The broker will implement a new component which will receive an activated job, its key, its variables, and the activation properties which were used to activate the job.
  - This is to be exposed to the engine, indirectly via an API of course.
- This component try pushing the job to all associated recipients for the given activation properties.
  - We can implement this in a very naive way initially, best-of effort well distributed, e.g. randomly or round-robin.
  - When a job fails to be pushed, the broker will retry the next possible recipient for the given activation properties.
  - If none of the recipients succeed, we can use a log stream writer to yield the job back to the engine.
  - If there was a leader change in between, the job can be considered "lost".
  - There is initially no flow control.

#### Job activation

- When a job would be made available, we should always try to activate it immediately if possible using the API described above. This means:
  - When a job fails, if there are retries left
  - When a job is created
  - When a job is retried with back off
  - When a job timed out
  - When a job-incident is resolved

**Comment by deepthidevaki** [comment]
An open question from https://github.com/camunda/zeebe/issues/11713#issuecomment-1482498625 
- Optimize retrying of request sent from client to server. Currently AddRequest and RemoveRequest are retried indefinitely. What is the impact of this? Should we stop retrying at some point? When we stop retrying, should we return an error to the client? Should we use backoff for retry?

**Comment by npepinpe** [comment]
Another open question:

- When pushing a job from the broker to the gateway times out, how should we handle this? We probably don't want to retry, as this could cause the job to be sent out twice. Do we want some form of handshake between broker/gateway before forwarding the job?

**Comment by deepthidevaki** [comment]
Added https://github.com/camunda/zeebe/issues/12663 as follow up.

**Comment by berkaycanbc** [comment]
A note from our meeting with @koevskinikola and @npepinpe.

In the initial phase, if pushing a job fails, we yield the job back and make it available (activatable) for long polling. In other words, initially, there will not be a mechanism to push the job again when it fails pushing.

**Comment by berkaycanbc** [comment]
A note from our meeting with @koevskinikola and @npepinpe.

In the initial phase, if pushing a job fails, we yield the job back and make it available (activatable) for long polling. In other words, initially, there will not be a mechanism to push the job again when it fails pushing.

**Comment by berkaycanbc** [comment]
A note from our meeting with @koevskinikola and @npepinpe.

In the initial phase, if pushing a job fails, we yield the job back and make it available (activatable) for long polling. In other words, initially, there will not be a mechanism to push the job again when it fails pushing.

**Comment by npepinpe** [comment]
Another open question, when yielding jobs back on failure, we could run into the following problems:

- Thundering herd, where we suddenly see a spike of failed jobs, resulting in a lot of commands being appended to the log at once. This leads to delays in further commands across the partition.
- Related, but yielding a job when the log is backed up could result in this job taking, say, 15 seconds to be yielded. If it will time out in 5, this is just pointless busy work.

**Comment by npepinpe** [comment]
Genau, :boom: 

We might want to include some form of sequencing or lease as part of the job yield, i.e.

- Activate job, returns you a "lease" (can be anything - logical clock, timestamp, etc.)
- On yield, you yield back that specific lease. If the current holder is not the same, it's a no-op

**Comment by koevskinikola** [comment]
I've included https://github.com/camunda/zeebe/issues/12773 in the tasks related to this ticket so we don't loose track of the issue.

ZPA will try to schedule time for it once work on the gateway/client side for this topic is done.

**Comment by npepinpe** [comment]
Another thought:

We may want to preemptively close long living gRPC streams once they reach a certain age to allow rebalancing between gateways and avoid having too many clients on the same gateway. This can be done in a future iteration, but it's likely that users will eventually run into the problem of having too many clients converging to the same gateways if they live long enough.

**Comment by npepinpe** [comment]
As we're nearing the end of the alpha scope, I would propose the following for the production scope:

- [ ] Implement flow control/workload distribution (see [Workload distribution](https://github.com/camunda/zeebe/issues/11231#issuecomment-1464101226))
- [ ] Improved error handling (see [Intelligent error handling](https://github.com/camunda/zeebe/issues/11231#issuecomment-1464101226))

And of course fixing any bugs and issues we find in the mean time :upside_down_face:

**Comment by deepthidevaki** [comment]
> Intelligent error handling

Don't we already retry with identical stream on both gateway and broker? What is missing?

**Comment by npepinpe** [comment]
Hm, I thought I had expanded that. The missing part (which is not in the linked comment, even if I was sure I'd added it :smile:) is about thundering herd issues. If we pushed out many jobs, and suddenly all fail (e.g. all time out), we would be yielding many jobs back and possibly overloading the writer/processor, leading to high back pressure.

**Comment by npepinpe** [comment]
One caveat to anyone thinking of using this before it's production ready. We already ran benchmarks with larger clusters, and without flow control, you not only have to scale your workers to accomodate the peak load at all times, but you have to also scale your gateways. There is no failsafe in the gateway to reject requests, so it will keep allocating memory to cope and will eventually crash.

When introducing flow control, we'll likely want to add two form of limiting: per client, and per gateway (since the gateway can only do so much anyway).

**Comment by korthout** [comment]
@npepinpe Do we have plans to support the stream activated jobs RPC in zeebe-process-test?

**Comment by npepinpe** [comment]
We should, although it's perfectly usable without.

**Comment by npepinpe** [comment]
All tasks were completed or pushed to the backlog, :tada:
